File: imunocare_clinic_ext/imunocare_clinic_ext/report/projeção_de_estoque_de_vacinas/projeção_de_estoque_de_vacinas.js
// Projeção de Estoque de Vacinas (Fase 11) — filtros, cores e largura total.
// Carregado em runtime pelo report view; não depende de build de assets.

frappe.query_reports["Projeção de Estoque de Vacinas"] = {
	filters: [
		{
			fieldname: "meses",
			label: __("Horizonte"),
			fieldtype: "Select",
			options: [
				{ value: "3", label: __("Próximos 3 meses") },
				{ value: "6", label: __("Próximos 6 meses") },
				{ value: "12", label: __("Próximos 12 meses") },
			],
			default: "3",
			reqd: 1,
		},
		{
			fieldname: "incluir_em_pedido",
			label: __("Considerar 'Em pedido' no saldo"),
			fieldtype: "Check",
			default: 1,
		},
		{
			fieldname: "somente_com_deficit",
			label: __("Só vacinas a repor ⚠"),
			fieldtype: "Check",
		},
	],

	formatter(value, row, column, data, default_formatter) {
		const formatted = default_formatter(value, row, column, data);
		const fn = column.fieldname || "";

		// Saldo mensal projetado: vermelho/negrito quando o estoque "fura".
		if (fn.startsWith("saldo_") && data) {
			const v = data[fn];
			if (v != null && v < 0) {
				return `<span style="color:var(--red-600,#c0392b);font-weight:700">${formatted}</span>`;
			}
			if (v != null && v <= 5) {
				return `<span style="color:var(--orange-600,#d35400);font-weight:600">${formatted}</span>`;
			}
			return formatted;
		}

		// Repor: pílula de alerta quando > 0 (quanto comprar para não faltar).
		if (fn === "repor" && data) {
			if (data.repor > 0) {
				return `<span class="indicator-pill red" style="font-weight:700"
					title="${__("Doses a repor no horizonte")}">⚠ ${formatted}</span>`;
			}
			return `<span class="indicator-pill green">${formatted}</span>`;
		}

		// Estoque atual zerado: destaque.
		if (fn === "estoque" && data && data.estoque <= 0) {
			return `<span style="color:var(--red-600,#c0392b);font-weight:700">${formatted}</span>`;
		}

		return formatted;
	},

	onload(report) {
		report.page.add_inner_button(__("Comprar (Material Request)"), () => {
			frappe.new_doc("Material Request", { material_request_type: "Purchase" });
		});

		// Largura total (mesma classe do toggle nativo do Frappe). Removida ao
		// sair da rota do report para não vazar para outras páginas.
		if (!document.body.classList.contains("full-width")) {
			document.body.classList.add("full-width");
			report._imun_added_fullwidth = true;
		}
		frappe.router.on("change", function _imun_restore() {
			if ((frappe.get_route_str() || "").indexOf("Proje") === -1) {
				if (report._imun_added_fullwidth) {
					document.body.classList.remove("full-width");
				}
			}
		});
	},
};

File: imunocare_clinic_ext/imunocare_clinic_ext/report/projeção_de_estoque_de_vacinas/projeção_de_estoque_de_vacinas.py
"""Report Projeção de Estoque de Vacinas (Fase 11).

Cruza a demanda dos agendamentos futuros (linhas Imunocare Appointment Vaccine,
1 linha = 1 dose) com a posição do estoque nativo (Bin) e projeta o saldo de
cada vacina mês a mês. Responde diretamente à pergunta "quanto preciso repor
para os próximos meses".

Por que não usar o nativo: Stock Projected Qty / Item Shortage / Production
Plan calculam projeção, mas a demanda deles vem só de Sales Order / Material
Request / Work Order — nunca de Patient Appointment. Toda a aritmética de
estoque, porém, reusa o Bin nativo (ver feedback_reuse_first).

Colunas dinâmicas: uma por mês do horizonte (saldo projetado no fim do mês),
geradas em runtime. A renderização (vermelho p/ saldo negativo, destaque do
"Repor", largura total) fica no ``.js`` do app — sem build de assets.
"""

from __future__ import annotations

from frappe import _

from imunocare_clinic_ext.api.dashboard import projetar_estoque


def execute(filters: dict | None = None):
	filters = filters or {}
	meses = int(filters.get("meses") or 3)
	incluir = filters.get("incluir_em_pedido")
	incluir = True if incluir is None else bool(incluir)
	so_deficit = bool(filters.get("somente_com_deficit"))

	proj = projetar_estoque(meses=meses, incluir_em_pedido=incluir)
	linhas = proj["linhas"]
	if so_deficit:
		linhas = [l for l in linhas if l["repor"] > 0]

	return _columns(proj["meses"]), _data(linhas)


def _columns(rotulos_mes: list[str]) -> list[dict]:
	cols = [
		{"label": _("Vacina"), "fieldname": "medication", "fieldtype": "Link", "options": "Medication", "width": 220},
		{"label": _("Estoque"), "fieldname": "estoque", "fieldtype": "Float", "precision": "0", "width": 90},
		{"label": _("Em pedido"), "fieldname": "em_pedido", "fieldtype": "Float", "precision": "0", "width": 95},
	]
	# Uma coluna por mês: saldo projetado no fim do mês (saldo_0, saldo_1, ...).
	for i, rotulo in enumerate(rotulos_mes):
		cols.append({"label": rotulo, "fieldname": f"saldo_{i}", "fieldtype": "Float", "precision": "0", "width": 90})
	cols += [
		{"label": _("Demanda"), "fieldname": "demanda_total", "fieldtype": "Float", "precision": "0", "width": 95},
		{"label": _("Repor"), "fieldname": "repor", "fieldtype": "Float", "precision": "0", "width": 100},
	]
	return cols


def _data(linhas: list[dict]) -> list[dict]:
	out = []
	for l in linhas:
		row = {
			"medication": l["medication"],
			"estoque": l["estoque"],
			"em_pedido": l["em_pedido"],
			"demanda_total": l["demanda_total"],
			"repor": l["repor"],
		}
		for i, saldo in enumerate(l["saldos"]):
			row[f"saldo_{i}"] = saldo
		out.append(row)
	return out
